### PR TITLE
190: Use option types rather than booleans for verification return signatures

### LINKED ISSUE #190: Use `Result` rather than `bool` for verification checks.
Currently, we use a mix of `Result`-valued and `bool`-valued verification checks.  This is somewhat dangerous, because it opens the possibility of bugs arising from the mismatch between the two styles, and inconvenient, because the `false` bools can't be wrapped with additional context like a `Result` can.  We should ensure that verification checks return a `Result` (even a `Result<(),()>`, though preferably a `Result<(), SomeError>`) rather than a `bool`.

### PR DESCRIPTION
**Pull request by zbuc**
Closes #190 

There are some cases where the deeper verification error is discarded and a generic error is bubbled up, I'm not sure if we want to address these or not but I commented them.

### CODE REVIEW

File: pd/src/verify.rs
Comment (redshiftzero) [inline pd/src/verify.rs:77]: As soon as we see an invalid proof in any action, we can bubble up the error as the entire transaction has failed

**Review by redshiftzero — APPROVED** [review]
LGTM, but I'll let @hdevalence give his thoughts on the FMD parts before merging

File: decaf377-fmd/src/detection.rs
Comment (hdevalence) [inline decaf377-fmd/src/detection.rs:81]: This method should return a `bool` rather than a `Result`, since (unlike, e.g., a signature verification), getting a `false` result doesn't indicate an error, but just that the clue was not intended for this detection key.

File: crypto/src/proofs/transparent.rs
Comment (hdevalence) [inline crypto/src/proofs/transparent.rs:36]: These are probably useful for debugging, but we could eventually replace this whole error enum with a single unit error type, since a snark verification will just succeed or fail, revealing nothing about the witness data.  (Maybe this replacement happens when we implement snark proofs, just noting it for later).

**Review by hdevalence — CHANGES_REQUESTED** [review]
This looks good except the FMD changes, noted in comment that the `examine` method returns a `bool` intentionally, but happy to merge this once that part is rolled back, other comments are just informational.

**Pull request merged**
